knora/dsplib/utils/onto_validate.py

```diff
@@ -38,15 +37,138 @@ def validate_ontology(input_file_or_json: Union[str, Dict, os.PathLike]) -> bool
 
     # validate the data model against the schema
     current_dir = os.path.dirname(os.path.realpath(__file__))
-
     with open(os.path.join(current_dir, '../schemas/ontology.json')) as s:
         schema = json.load(s)
-
     try:
-        validate(instance=data_model, schema=schema)
+        jsonschema.validate(instance=data_model, schema=schema)
     except jsonschema.exceptions.ValidationError as err:
         print(f'Data model did not pass validation. The error message is: {err.message}\n'
               f'The error occurred at {err.json_path}')
         return False
-    print('Data model is syntactically correct and passed validation.')
-    return True
+
+    # cardinalities check for circular references
+    if check_cardinalities_of_circular_references(data_model):
+        print('Data model is syntactically correct and passed validation.')
+        return True
+    else:
+        return False
+
+
+def check_cardinalities_of_circular_references(data_model: dict[Any, Any]) -> bool:
+    """
+    Check if there are properties derived from hasLinkTo that form a circular reference. If so, these
+    properties must have the cardinality 0-1 or 0-n, because during the xmlupload process, these values
+    are temporarily removed.
+    """
+
+    # search the ontology for all properties that are derived from hasLinkTo. store them in a dict, and map
+    # them to their objects (i.e. the resource classes they point to)
+    # example: if the property 'rosetta:hasTextMedium' points to 'rosetta:Image2D':
+    # link_properties = {'rosetta:hasTextMedium': ['rosetta:Image2D'], ...}
+    ontos = data_model['project']['ontologies']
```

Comment:
Theoretically you could create this dict after identifying problematic cases so this dict doesn't get as large. Can't imagine this becoming a problem though.

Reply:
In order to identify the problematic cases, I need this dict first.